Trait coherence: disallow inherent impls on external types

tritao opened the pull request:
This improves the trait coherence checks for impl self case, and polishes a bit of the wording on the existing diagnostics.

We now enforce package-level check in impl-self type checking, which means we now reject inherent impls for external nominal types (struct/enum).

There is still is a specialized whitelist for `StorageKey` so current code that uses this pattern keeps working.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enforces package-level coherence by rejecting inherent `impl` blocks for external nominal types (with a whitelist for `std::storage::StorageKey<_>`), adds a dedicated diagnostic, and updates tests and orphan rule wording from module to package.
> 
> - **Semantic analysis**:
>   - Disallow inherent `impl` for external nominal types in `type_check_impl_self` (`sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs`).
>   - Temporary whitelist for `std::storage::StorageKey<_>`.
>   - Emit new error `CompileError::InherentImplForExternalType` with spans.
> - **Diagnostics** (`sway-error/src/error.rs`):
>   - Add `InherentImplForExternalType` variant and full diagnostic with hints/help.
>   - Change orphan rule wording from "module" to "package" in messages and help.
> - **Tests**:
>   - Update failing snapshots and checks to expect package-level wording and the new inherent-impl error.
>   - Refactor `should_pass/language/generic_transpose` to use a new trait `OptionTranspose` instead of an inherent impl on `Option<_>`.
>   - Adjust `impl_self_overlap` and orphan rule snapshots accordingly.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 308d0971d41bf70027c18f64453b9e9d4471a723. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

tritao commented:
Rebased due to conflicts.

ironcev commented on sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs:
```diff
@@ -527,6 +527,56 @@ impl TyImplSelfOrTrait {
+                        match pkg_name {
+                            Some(name) => name != current_pkg.name().as_str(),
+                            None => false,
+                        }
```
Same here.
```suggestion
                        pkg_name.and_then(|name| name != current_pkg.name().as_str()).unwrap_or_default()
```

ironcev commented on sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs:
```diff
@@ -527,6 +527,56 @@ impl TyImplSelfOrTrait {
+
+                if is_external && !is_storage_key_in_std {
+                    let type_name = engines.help_out(implementing_for.type_id).to_string();
+                    let type_definition_span = match &*type_engine.get(implementing_for.type_id) {
```
Also here we should handle aliases. Ideally, we could have a diagnostic saying that the declaration type is an alias for the external type. But for now we can just make sure we handle this.
```suggestion
                    let type_definition_span = match &*type_engine.get_unaliased(implementing_for.type_id) {
```

tritao committed:
Trait coherence: disallow inherent impls on external types

This improves the trait coherence checks for impl self case,
and polishes a bit of the wording on the existing diagnostics.

* Add CompileError::InherentImplForExternalType and diagnostics.
  Reason/issue/help now refer to "package" to match coherence scope

* Enforce package-level check in impl-self type checking
  Reject inherent impls for external nominal types (struct/enum)

Temporary whitelist: allow inherent impls on std::storage::StorageKey<_>

This is a workaround so current code that uses this pattern keeps
working.

ironcev committed:
Merge branch 'master' into improve-trait-coherence-checks
